### PR TITLE
Count the whole softmax family and stop dividing by an empty axis

### PR DESCRIPTION
**Pull request by raimbekovm**
## summary

`nn.LogSoftmax`, `nn.Softmin` and `nn.Softmax2d` reached no counting rule, so each was charged zero MACs while doing the same exponentiate, sum and divide over the same tensor that `nn.Softmax` is counted for. They share no private base, so each member is named.

All three then differ from a plain softmax by one term, and `count_softmax` reads each off the module rather than needing a rule apiece:

- `nn.Softmax2d` carries no `dim` attribute at all. It always normalizes the channel axis, which is the third from the end for both the batched and the unbatched input it accepts.
- `nn.Softmin` runs as `(-x).softmax(dim)`, so it negates every element first — one operation per element.
- `nn.LogSoftmax` runs as `x - logsumexp(x)`, so it takes one logarithm per normalized vector.

Both extra terms are worth writing because `calculate_softmax` charges `nfeatures - 1` additions rather than `nfeatures`: the formula is already exact to within one operation per vector, which is the size of the smaller of the two.

The resolution for a `dim=None` softmax is unchanged and still runs for the three classes that do have the attribute.

Separately, `count_softmax` derived the number of normalized vectors by dividing the element count by the axis length, which raised `ZeroDivisionError` when that axis was empty. torch runs the layer, so a model that works could not be profiled:

```python
nn.Softmax(dim=1)(torch.zeros(2, 0))  # torch.Size([2, 0])
profile(nn.Sequential(nn.Softmax(dim=1)), inputs=(torch.zeros(2, 0),), verbose=False)
```

An empty axis is no work at all and leaves no vector count to divide out, so it now reports 0.

No published figure moves. Neither the README benchmark table nor any torchvision model instantiates one of these four as a module, a fifteen-model torchvision differential is byte-identical, and both README worked examples still produce the values they print.

### Measured impact

Over an `8x1000` input, with `nn.Softmax` alongside as the member that was already counted:

| layer                            | before | after   |
| -------------------------------- | ------ | ------- |
| `nn.Softmax(dim=1)`              | 23,992 | 23,992  |
| `nn.LogSoftmax(dim=1)`           | 0      | 24,000  |
| `nn.Softmin(dim=1)`              | 0      | 31,992  |
| `nn.Softmax2d` over `1x64x32x32` | 0      | 195,584 |

Each of the three lands where its one extra term puts it: `nn.LogSoftmax` 8 above the plain softmax, one logarithm per normalized vector, and `nn.Softmin` 8,000 above it, one negation per element, +33%.

`nn.Softmax(dim=1)` over a `(2, 0)` input raised `ZeroDivisionError` and now reports 0.

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary

🧮 Adds operation-counting support for PyTorch `LogSoftmax`, `Softmin`, and `Softmax2d` layers with improved edge-case handling.

### 📊 Key Changes

- Registers `nn.LogSoftmax`, `nn.Softmin`, and `nn.Softmax2d` with the existing softmax counting hook.
- Updates `count_softmax` to:
  - Handle `Softmax2d` normalization across channels, including inputs with or without a batch dimension.
  - Preserve PyTorch’s behavior for deprecated implicit dimension selection.
  - Safely process scalar and zero-feature inputs.
  - Add operation costs specific to `Softmin` and `LogSoftmax`.

### 🎯 Purpose & Impact

- ✅ Produces more accurate FLOP and operation estimates for models using the full PyTorch softmax family.
- 🛡️ Improves robustness for scalar, empty, and unusual input shapes.
- 📈 Helps profiling tools better reflect the computational cost of classification and normalization layers without affecting existing supported layers.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @raimbekovm, thank you for submitting a `ultralytics/thop` 🚀 PR! This automated message confirms your contribution has been received, and an engineer will assist with review. To ensure a seamless integration of your work, please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/thop/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/thop` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The Softmax2d axis handling and empty-axis guard look sound. However, Softmin is now systematically undercounted because its per-element negation is omitted.

💬 Posted 1 inline comment

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The added rules correctly cover the previously unregistered softmax module variants, and the zero-length-axis guard prevents the division-by-zero failure. One counting-accuracy gap remains for `LogSoftmax`: its per-vector logarithm is not included.

💬 Posted 1 inline comment

**Review by UltralyticsAssistant — APPROVED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Clean PR. The new registrations resolve all named softmax module variants, the Softmax2d axis handling matches its supported tensor layouts, and the empty-axis guard prevents division by zero while preserving zero work for empty tensors.

**Review by glenn-jocher — APPROVED** [review]
LGTM — cold-reviewed the full live-head diff against the Core Principles, package performance, the supported Python/PyTorch floor, and the combined eight-PR state. No findings.

**Review by glenn-jocher — APPROVED** [review]
LGTM on the updated exact head. The PR diff remains the reviewed 14-line softmax-family owner change after merged #150–#152, and all checks are green.

**Comment by UltralyticsAssistant** [comment]
🎉 Merged in [PR #153](https://github.com/ultralytics/thop/pull/153)!

Thank you @raimbekovm, with contributions from @glenn-jocher, for expanding operation-counting support across PyTorch’s `LogSoftmax`, `Softmin`, and `Softmax2d` layers. The improved handling of batch-free, scalar, empty, and deprecated-dimension cases makes THOP profiling more accurate and robust for real-world models.

> “Great things are done by a series of small things brought together.” — Vincent van Gogh

This PR is a great example: careful edge-case support and layer coverage combine to give the community more dependable FLOP and operation estimates. We appreciate your work! 🚀
